libusb/libusb.h

```diff
@@ -953,6 +957,56 @@ struct libusb_ss_usb_device_capability_descriptor {
 	uint16_t bU2DevExitLat;
 };
 
+/** \ingroup libusb_desc
+ * A enum to extract attributes from
+ * \ref libusb_ssplus_usb_device_capability_descriptor.bmAttributes
+ */
+enum libusb_ssplus_attribute_mask {
+	LIBUSB_SSPLUS_ATTRIBUTE_SSCA_MASK = 0xF,
+	LIBUSB_SSPLUS_ATTRIBUTE_SSIC_MASK = 0xF0
+};
+
+/** \ingroup libusb_desc
+ * A structure representing the SuperSpeedPlus descriptor
+ * This descriptor is documented in section 9.6.2.5 of the USB 3.1 specification.
+ */
+struct libusb_ssplus_usb_device_capability_descriptor {
+	/** Size of this descriptor (in bytes) */
+	uint8_t  bLength;
+	
+	/** Descriptor type. Will have value
+	 * \ref libusb_descriptor_type::LIBUSB_DT_DEVICE_CAPABILITY
+	 * LIBUSB_DT_DEVICE_CAPABILITY in this context. */
+	uint8_t  bDescriptorType;
+	
+	/** Capability type. Will have value
+	 * \ref libusb_capability_type::LIBUSB_BT_SUPERSPEED_PLUS_CAPABILITY
+	 * LIBUSB_BT_SUPERSPEED_PLUS_CAPABILITY in this context. */
+	uint8_t  bDevCapabilityType;
+	
+	/** Reserved field */
+	uint8_t  bReserved;
+	
+	/** Bitmap encoding of:
+	 * ssac: (sublink speed attributes number)
+	 * ssic: (sublink speed id number)
+	 * See enum libusb_ssplus_attribute_mask for convenient masks*/
+	uint32_t bmAttributes;
+	
+	/** Bitmap encoding of supported device level functionalities.
+	 *  ssid, min RX lanes, min TX lanes.
+	 */
+	uint16_t wFunctionalitySupport;
+	
+	/** Reserved field */
+	uint16_t wReserved;
+	
+	/** An array of sublink speed attributes. See \ref bmAttributes for
+	 * the number of attributes  (ssid+1).
```

Comment:
Could you please improve the comments about the length of this array? I've seen \ref bmAttributes as instructed, but it does not mention anything named `ssid`...

Reply:
Sorry about this typo. I mean `ssic`, not `ssid`.

Reply:
Thanks!

So the length of `bmSublinkSpeedAttr` is `(bmAttributes & LIBUSB_SSPLUS_ATTRIBUTE_SSIC_MASK) + 1`?

Could you please put that explicitly also?  Buffer overruns are all too easy in C, so best to make it *absolutely clear* what the length of arrays are. Your comment is good, but it did require scrolling up to "See \ref bmAttributes", then from there scrolling up to find the bit mask, then scrolling down to remember the + 1.  Even then, I'm not sure I'm right about the length...

Reply:
Yes you got the length right. I don't think I can make it more explicit than what is currently stated (`ssid+1`).

Reply:
It is pretty clear, and you have written very nice comments in this PR!

But buffer overruns are ***so easy*** and ***so common*** in C that I think *even more* clarity would help. Concretely:

```diff
-   /** An array of sublink speed attributes. See \ref bmAttributes for
-    * the number of attributes  (ssic+1).
-    */
-   uint32_t* bmSublinkSpeedAttr;
+   /** An array of sublink speed attributes. See \ref bmAttributes for
+    * the number of attributes (ssic+1). Concretely, this array has
+    * (bmAttributes & LIBUSB_SSPLUS_ATTRIBUTE_SSIC_MASK) + 1 elements.
+    */
+   uint32_t* bmSublinkSpeedAttr;
```

Now let's go check how you loop over this array... Hmmm, you have:

```c
for(i=0 ; i < (ssplus_usb_device_cap->bmAttributes & LIBUSB_SSPLUS_ATTRIBUTE_SSCA_MASK) + 1 ; i++) {
	printf("            SubLink capability: %04X\n", ssplus_usb_device_cap->bmSublinkSpeedAttr[i]);
}
```

Looks like you used the wrong mask!

Reply:
No the mask was write, I messed up on the comment. The mask is indeed ssca. I have fixed the comment.